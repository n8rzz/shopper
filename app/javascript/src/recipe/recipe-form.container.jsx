import React from 'react';
import propTypes from 'prop-types';

export function RecipeFormContainer(props) {
    return (
        <div>RecipeFormContainer</div>
    );
}

RecipeFormContainer.propTypes = {

};
