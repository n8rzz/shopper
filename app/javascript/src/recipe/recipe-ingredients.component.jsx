import React from 'react';
import PropTypes from 'prop-types';

export function RecipeIngredientsComponent(props) {
    const _buildRecipeIngredientsJsx = () => props.ingredients.map((ingredient, index) => {
        const itemName = ingredient.name || props.itemsById[ingredient.item_id].name;

        return (
            <li
                key={`ingredient-${index}`}
                className={'recipeIngredients-list-item'}
            >
                <ul className={'hlist'}>
                    <li>{ ingredient.qty_value } <span>{ ingredient.qty_unit }</span></li>
                    <li>{ itemName }</li>
                </ul>
            </li>
        );
    });

    if (props.ingredients.length === 0) {
        return null;
    }

    return (
        <div className={'recipeIngredients'}>
            <ul className={'recipeIngredients-list'}>
                { _buildRecipeIngredientsJsx() }
            </ul>
        </div>
    );
}

RecipeIngredientsComponent.propTypes = {
    ingredients: PropTypes.array.isRequired,
    itemsById: PropTypes.object.isRequired,
};
