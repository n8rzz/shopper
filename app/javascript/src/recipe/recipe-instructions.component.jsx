import React from 'react';
import PropTypes from 'prop-types';

export function RecipeInstructionsComponent(props) {
    if (props.instructions.length === 0) {
        return null;
    }

    return (
        <div className={'recipeInstructions'}>
            <ol className={'recipeInstructions-list'}>
                {props.instructions.map((instruction, index) => (
                    <li
                        key={`instruction-${index}`}
                        className={'recipeInstructions-list-item'}
                    >
                        {instruction.text}
                    </li>
                ))}
            </ol>
        </div>
    );
}

RecipeInstructionsComponent.propTypes = {
    instructions: PropTypes.array.isRequired,
};
