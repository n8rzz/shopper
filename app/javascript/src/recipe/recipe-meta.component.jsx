import React from 'react';
import PropTypes from 'prop-types';

export function RecipeMetaComponent(props) {
    const _buildRecipeMetaJsx = () => (
        <ul className={'recipeMeta-list'}>
            <li className={'recipeMeta-list-item'}>
                <div>Prep</div>
                <div>{ props.prepTime } minutes</div>
            </li>
            <li className={'recipeMeta-list-item'}>
                <div>Cook</div>
                <div>{ props.cookTime } minutes</div>
            </li>
            <li className={'recipeMeta-list-item'}>
                <div>Serves</div>
                <div>{`${props.yieldValue} ${props.yieldUnit}`}</div>
            </li>
        </ul>
    );

    return (
        <div className={'recipeMeta'}>
            <div className={'vr_4'}>
                { props.description }
            </div>
            { _buildRecipeMetaJsx() }
        </div>
    );
}

RecipeMetaComponent.propTypes = {
    description: PropTypes.string.isRequired,
    prepTime: PropTypes.number.isRequired,
    cookTime: PropTypes.number.isRequired,
    yieldValue: PropTypes.number.isRequired,
    yieldUnit: PropTypes.string.isRequired,
};
