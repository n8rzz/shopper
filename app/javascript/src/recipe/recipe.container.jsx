import React from 'react';
import PropTypes from 'prop-types';
import { RecipeIngredientsComponent } from './recipe-ingredients.component';
import { RecipeInstructionsComponent } from './recipe-instructions.component';
import { RecipeMetaComponent } from './recipe-meta.component';

export function RecipeContainer(props) {
    // TODO: maybe .vlist instead of .vr?
    return (
        <ul className={'vlist'}>
            <li>
                <RecipeMetaComponent
                    description={props.recipe.description}
                    prepTime={props.recipe.prep_time}
                    cookTime={props.recipe.cook_time}
                    yieldValue={props.recipe.yield_value}
                    yieldUnit={props.recipe.yield_unit}
                />
            </li>
            <li>
                <RecipeIngredientsComponent
                    ingredients={props.ingredients}
                    itemsById={props.itemsById}
                />
            </li>
            <li>
                <RecipeInstructionsComponent instructions={props.instructions} />
            </li>
        </ul>
    );
}

RecipeContainer.propTypes = {
    recipe: PropTypes.object.isRequired,
    ingredients: PropTypes.array.isRequired,
    instructions: PropTypes.array.isRequired,
    itemsById: PropTypes.object.isRequired,
};
