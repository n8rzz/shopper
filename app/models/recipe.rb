# frozen_string_literal: true

# name: string
# description: string
# prep_time: integer
# cook_time: integer
# yield_value: string
# yield_unit: string
# ownable: references
# assembly: references
class Recipe < ApplicationRecord
  belongs_to :assembly, optional: true
  belongs_to :ownable, polymorphic: true

  has_many :recipe_ingredients
  has_many :recipe_instructions

  accepts_nested_attributes_for :recipe_ingredients, allow_destroy: true
  accepts_nested_attributes_for :recipe_instructions, allow_destroy: true

  validates :name, uniqueness: true
  validate :name_xor_assembly_reference

  def display_name
    return name if assembly.nil?

    assembly.name
  end

  private

  def name_xor_assembly_reference
    unless name.blank? ^ assembly.blank?
      errors[:base] << 'Specify an assembly_id or a Name but not both'
    end
  end
end
