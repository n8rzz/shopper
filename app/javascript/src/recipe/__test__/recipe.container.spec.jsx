import React from 'react';
import { shallow } from 'enzyme';
import {
    recipeIngredientMock,
    recipeInstructionMock,
    recipeMock,
} from '../../__mocks__/mocks';
import { RecipeContainer } from '../recipe.container';

describe('RecipeContainer', () => {
    test('renders correctly with valid props', () => {
        const component = shallow(<RecipeContainer
            recipe={recipeMock}
            ingredients={[recipeIngredientMock]}
            instructions={[recipeInstructionMock]}
            itemsById={{}}
        />);

        expect(component).toMatchSnapshot();
    });
});
