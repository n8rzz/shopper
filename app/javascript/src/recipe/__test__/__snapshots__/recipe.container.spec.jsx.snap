// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeContainer renders correctly with valid props 1`] = `
<ul
  className="vlist"
>
  <li>
    <RecipeMetaComponent
      cookTime={18}
      description="Two buttermilk waffles, topped with whipped cream and maple syrup, a side of two eggs served any style, and your choice of smoked bacon or smoked ham."
      prepTime={14}
      yieldUnit="servings"
      yieldValue={28}
    />
  </li>
  <li>
    <RecipeIngredientsComponent
      ingredients={
        Array [
          Object {
            "item_id": null,
            "name": "Basil",
            "qty_unit": "Monkeys",
            "qty_value": "10",
            "recipe_id": 1,
          },
        ]
      }
      itemsById={Object {}}
    />
  </li>
  <li>
    <RecipeInstructionsComponent
      instructions={
        Array [
          Object {
            "recipe_id": 1,
            "sort_order": 0,
            "text": "Migas normcore heirloom pug jean shorts gentrify irony. Typewriter flexitarian synth dreamcatcher tilde swag salvia. Flannel paleo taxidermy viral muggle magic. Banjo roof cleanse semiotics.",
          },
        ]
      }
    />
  </li>
</ul>
`;
